src/rotation_types.jl

```diff
@@ -0,0 +1,291 @@
+import Base: convert, call, getindex, *, eltype
+
+
+numel{T}(::Type{T}) = length(fieldnames(T))  # for getting the number of elements in the representation
+                                             # e.g. length(Quaternion) == 1                # its a number
+                                             # e.g. length(fieldnames(Quaternion)) == 5    # it has a normalized field
+                                             # e.g. numel(Quaternion) == 4                 # what we want
+
+
+# build a dictionary to build transformation paths from one representation to the next
+# I hate this but trying to make a generic overload for convert causes no end of problems...
+trans_dict = Dict()
+
+
+###################################################
+# Rotation matrix (typeofalias of Mat{3,3,Float64})
+# N.B. all other parameterizations T should implement
+# a convert_rotation(T, RotMatrix) method
+###################################################
+
+"""
+A type alias for rotation matrices (alias of Mat{3,3,T})
+"""
+typealias RotMatrix{T <: Real}   Mat{3,3,T}
+
+# Fixes for FixedSizeArray stuff
+@inline convert{T}(::Type{RotMatrix{T}}, X::RotMatrix{T}) = X                   # without this the conversion isn't type safe
+@inline convert{U}(::Type{RotMatrix}, mat::Matrix{U}) = RotMatrix{U}(mat)       # this hangs othewise
+@inline call{U}(::Type{RotMatrix}, mat::Matrix{U}) = convert(RotMatrix{U}, mat) # this hangs othewise
+
+# allow a RotMatrix to be constructructed from 9 elements
+@inline call(::Type{RotMatrix}, x1::Real, x2::Real, x3::Real, x4::Real, x5::Real, x6::Real, x7::Real, x8::Real, x9::Real) = RotMatrix((x1,x2,x3), (x4,x5,x6), (x7,x8,x9))
+@inline call{T1,T2,T3}(::Type{RotMatrix}, c1::NTuple{3,T1}, c2::NTuple{3,T2}, c3::NTuple{3,T3}) = RotMatrix{promote_type(T1,T2,T3)}(c1, c2, c3)
```

Comment:
I get:

```
julia> RotMatrix((1.0, 0.0, 0.0), (0.0, 1.0, 0.0), (0.0,0.0,1.0))
FixedSizeArrays.Mat{3,3,Tuple{Float64,Float64,Float64}}(
    (1.0,0.0,0.0) (1.0,0.0,0.0) (1.0,0.0,0.0)
    (0.0,1.0,0.0) (0.0,1.0,0.0) (0.0,1.0,0.0)
    (0.0,0.0,1.0) (0.0,0.0,1.0) (0.0,0.0,1.0)
```

with FixedSizeArrays v"0.1.0" and master


On 10 May 2016 at 15:14, Chris Foster <notifications@github.com> wrote:

> In src/rotation_types.jl
> <https://github.com/FugroRoames/Rotations.jl/pull/2#discussion_r62613369>:
>
> > +# a convert_rotation(T, RotMatrix) method
> > +###################################################
> > +
> > +"""
> > +A type alias for rotation matrices (alias of Mat{3,3,T})
> > +"""
> > +typealias RotMatrix{T <: Real}   Mat{3,3,T}
> > +
> > +# Fixes for FixedSizeArray stuff
> > +@inline convert{T}(::Type{RotMatrix{T}}, X::RotMatrix{T}) = X                   # without this the conversion isn't type safe
> > +@inline convert{U}(::Type{RotMatrix}, mat::Matrix{U}) = RotMatrix{U}(mat)       # this hangs othewise
> > +@inline call{U}(::Type{RotMatrix}, mat::Matrix{U}) = convert(RotMatrix{U}, mat) # this hangs othewise
> > +
> > +# allow a RotMatrix to be constructructed from 9 elements
> > +@inline call(::Type{RotMatrix}, x1::Real, x2::Real, x3::Real, x4::Real, x5::Real, x6::Real, x7::Real, x8::Real, x9::Real) = RotMatrix((x1,x2,x3), (x4,x5,x6), (x7,x8,x9))
> > +@inline call{T1,T2,T3}(::Type{RotMatrix}, c1::NTuple{3,T1}, c2::NTuple{3,T2}, c3::NTuple{3,T3}) = RotMatrix{promote_type(T1,T2,T3)}(c1, c2, c3)
>
> Why the version with three columns of tuples?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/FugroRoames/Rotations.jl/pull/2/files/fb3a7882257b93afed97cfaa9bf56669bbf410dc#r62613369>
>